kasa/klapprotocol.py

```diff
@@ -0,0 +1,512 @@
+"""Implementation of the TP-Link Smart Home Protocol.
+
+Comment by sdb - 4-Jul-2023
+
+Encryption/Decryption methods based on the works of
+Simon Wilkinson and Chris Weeldon
+
+While working on these changes I discovered my HS100 devices would periodically change their device owner to something that produces the following
+md5 owner hash: 994661e5222b8e5e3e1d90e73a322315.  It seems to be after an update to the on/off state that was scheduled via the app.  Switching the device on and off manually via the
+Kasa app would revert to the correct owner.
+
+For devices that have not been connected to the kasa cloud the theory is that blank username and password md5 hashes will succesfully authenticate but
+at this point I have been unable to verify.
+
+https://gist.github.com/chriswheeldon/3b17d974db3817613c69191c0480fe55
+https://github.com/python-kasa/python-kasa/pull/117
+
+N.B. chrisweeldon implementation had a bug in the encryption logic for determining the initial seq number and Simon Wilkinson's implementation did not seem to support
+incrementing the sequence number for subsequent encryption requests
+
+"""
+import asyncio
+import binascii
+import datetime
+import hashlib
+import logging
+from pprint import pformat as pf
+from typing import Any, Dict, Optional, Union
+
+import aiohttp
+
+# pycryptodome
+from Crypto import Random
+from Crypto.Cipher import AES
+from Crypto.Util import Padding
+from yarl import URL
+
+from .auth import AuthCredentials, TPLinkAuthProtocol
+from .exceptions import SmartDeviceAuthenticationException, SmartDeviceException
+from .json import dumps as json_dumps
+from .json import loads as json_loads
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TPLinkKlap(TPLinkAuthProtocol):
+    """Implementation of the KLAP encryption protocol.
+
+    KLAP is the name used in device discovery for TP-Link's new encryption
+    protocol, used by newer firmware versions.
+    """
+
+    DEFAULT_PORT = 80
+    DISCOVERY_PORT = 20002
+    DISCOVERY_TARGET = ("255,255,255,255", DISCOVERY_PORT)
+    DISCOVERY_BROADCAST_PAYLOAD = binascii.unhexlify("020000010000000000000000463cb5d3")
+    DISCOVERY_QUERY = {"system": {"get_sysinfo": None}}
```

Comment:
Let's keep the discovery parts inside the `discover.py`. I bought a tapo device and looks like it also supports this discovery mode, so we could inform users about such devices and advice them to look for using tapo library for comms:
```
DEBUG:kasa.discover:Got {'error_code': 0,
 'result': {'device_id': 'xxxx',
            'device_model': 'P110(EU)',
            'device_type': 'SMART.TAPOPLUG',
            'factory_default': False,
            'ip': '192.168.xxx',
            'is_support_iot_cloud': True,
            'mac': '48-22-54-xx',
            'mgt_encrypt_schm': {'encrypt_type': 'AES',
                                 'http_port': 80,
                                 'is_support_https': False,
                                 'lv': 2},
            'obd_src': 'tplink',
            'owner': 'xxxx'}}
```

The handshake itself looks different (https://github.com/fishbigger/TapoP100/blob/main/PyP100/PyP100.py#L113) and I'm just getting this as a response for `/app/handshake1` when trying to query the device:
```
HTTP/1.1 200 OK
Server: SHIP 2.0
Content-Length: 49
Content-Type: text/html

<html><body><center>200 OK</center></body></html>
```

Reply:
This should be ok now